internal/beater/api/telemetry/handler.go

```diff
@@ -0,0 +1,77 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package telemetry
+
+import (
+	"github.com/elastic/apm-server/internal/beater/request"
+	"github.com/elastic/elastic-agent-libs/monitoring"
+	"io"
+	"net/http"
+)
+
+var (
+	// MonitoringMap holds a mapping for request.IDs to monitoring counters
+	MonitoringMap = request.DefaultMonitoringMapForRegistry(registry)
+	registry      = monitoring.Default.NewRegistry("apm-server.telemetry")
+)
+
+type HandlerConfig struct {
+	// TelemetryUrl contains the URL to send telemetry to
+	TelemetryUrl string
+
+	// ClusterId contains the elastic cloud cluster ID
+	ClusterId string
+
+	// Version contains the APM Server version
+	Version string
+}
+
+func Handler(cfg HandlerConfig) request.Handler {
+
+	return func(c *request.Context) {
+		err := sendTelemetry(cfg.TelemetryUrl, c.Request.Header.Get("content-type"), c.Request.Body, cfg.ClusterId, cfg.Version)
```

Comment:
`/!\` the current issue is around here, we can properly get the `content-length` header with the expected value, but the original request body appears to be empty :-(

Reply:
After banging my head against this much longer than I'd like to admit (why is `ContentLength` -1, even when there's a valid body??? I *still* don't know) -- I think the problem was that you were trying to pass the incoming request's `io.Reader` to the outbound request without repackaging it. Not certain why that would be a problem, since you weren't calling `ReadAll` on it, but with my contributions where I read it out to a `[]byte` and then create a new buffer wrapped in an `io.Reader`, it's working for me. Perhaps it gets closed out of band or something? I am not sure.

Reply:
Thanks for taking care of this, I also tried doing a copy, but without any success, so congrats for finding a work-around !